src/vs/workbench/api/node/mainThreadLanguageFeatures.ts

```diff
@@ -137,13 +140,24 @@ export class MainThreadLanguageFeatures extends MainThreadLanguageFeaturesShape
 
 	$registerReferenceSupport(handle: number, selector: vscode.DocumentSelector): TPromise<any> {
 		this._registrations[handle] = modes.ReferenceProviderRegistry.register(selector, <modes.ReferenceProvider>{
-			provideReferences: (model: IReadOnlyModel, position: EditorPosition, context: modes.ReferenceContext, token: CancellationToken): Thenable<modes.Location[]> => {
-				return wireCancellationToken(token, this._proxy.$provideReferences(handle, model.uri, position, context));
+			provideReferences: (model: IReadOnlyModel, position: EditorPosition, context: modes.ReferenceContext, progress: (value: modes.Location) => void, token: CancellationToken): Thenable<modes.Location[]> => {
+				const progressHandle = this._progressRegistrations.registerChild(handle, progress);
+				let refs = this._proxy.$provideReferences(handle, progressHandle, model.uri, position, context);
+				refs = always(refs, () => this._progressRegistrations.unregisterChild(handle, progressHandle));
+				return wireCancellationToken(token, refs);
 			}
 		});
 		return undefined;
 	}
 
+	$provideReferenceProviderProgress(handle: number, progressHandle: number, location: modes.Location): TPromise<any> {
```

Comment:
This is unlike the other functions here, and is the first of its kind. For this reason, I had the TODO in the interface declaration to come up with a better name. It is more like a `provide...` than a `register...` in the sense that it is sending data to the host from the extension. The difference is that `provide...` methods are called by the main thread whereas this method is called by the extension to send progress to the main thread.

I think I will go ahead an give this a unique prefix like `notify` and add documentation to the interface.